compiler-core/templates/prelude.js

```diff
@@ -116,6 +116,9 @@ export class BitString {
   }
 
   constructor(buffer) {
+    if (buffer instanceof ArrayBuffer) {
+      buffer = new Uint8Array(buffer);
+    }
```

Comment:
Nice!

Looking at the code I don't understand why it is that it doesn't work with an array buffer. Did you work out why? Perhaps it should always be an array buffer.

How about we throw an exception if it's not the desired type? Someone could make the same mistake with one of the other buffer classes.

Reply:
Yeah I may have been a bit hasty to make a pull request, there may be a better way here or this may not even really require a fix. 

My understanding is that a javascript `ArrayBuffer` is an "array of bytes" ([docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer)), i.e. it can never be a true `BitString`. 

In the context of the issue, I was confused because I was trying to write a `BitString`'s contents to a file, and then read the content's back out, and they were failing an equality check. I think this was because my `BitString` literal used the `:utf8` option, so I think the constructed `BitString` must have been wrapped in a `Uint8Array` view, which is why the same data failed an equality check.

My solution was to cast the contents read out with the same `Uint8Array` data view. When working on this, since the buffer will never not be bytes, it made sense to just set a default for the 90% use case. If you need a different view you can always `const uint16Array = new Uint16Array(uint8Array.buffer)`. 

But now I'm thinking it may be more consistent to throw an exception on anything that isn't an `ArrayBuffer` as you suggest. I think I'll try it an see if some errors pop up regarding the construction of bitstrings with utf8 encodings, and try and get a deeper understanding of the issue.

Reply:
That sounds fab! Thank you.

Reply:
Are you still working on this @bcpeinhardt ? :)

Reply:
Yes :)
Sorry it's been kind of a crazy month for me, I can't believe it's been 20 days since I was looking at this 😂
I'm going to try to finish up the issues I've been working on this weekend 😌